### PR TITLE
[NHWC] Enable batch norm by refactoring OCL kernel

### PR DESCRIPTION
**Pull request by ce1adon**

### CODE REVIEW

**Review by atamazov — DISMISSED** [review]
Some cosmetic changes + recommendations

UPDATE: "Cosmetic" changes are actually mandatory.

File: src/kernels/MIOpenBatchNormBwdSpatial.cl [hunk 1]
```diff
@@ -396,19 +416,47 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
...
+    _FLOAT xread;
+    _FLOAT_PREC xhat;
+#else
     _FLOAT4 dyRead4;
     _FLOAT4 xread4;
     _FLOAT_PREC4 xhat4;
+#endif
 #if(MIO_BN_N > MIO_BN_LOOP_UNROLL_MAXN)
-    __attribute__((opencl_unroll_hint(4))) for(unsigned int k = lid << 2; k < MIO_BN_LESS4;
+    __attribute__((opencl_unroll_hint(4))) for(unsigned int k =
```
[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl:429]:
```suggestion
#if MIO_LAYOUT_NHWC == 1
    unsigned int k = lid;
#else
    unsigned int k = lid << 2;
#endif
    __attribute__((opencl_unroll_hint(4))) for(; k < MIO_BN_LESS4; k += GRPRD)
```

[the remark this answers is not in the document]
Comment (DrizztDoUrden) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl:429]: Maybe even `unsigned int k = lid << (2 * MIO_LAYOUT_NHWC);` if that compiles good and MIO_LAYOUT_NHWC is guaranteed to be either 0 or 1.

File: src/kernels/MIOpenBatchNormFwdTrainSpatial.cl [hunk 2]
```diff
@@ -329,13 +350,23 @@ MIOpenBatchNormFwdTrainSpatial(const __global _FLOAT* __restrict in,
...
-    __attribute__((opencl_unroll_hint(2))) for(unsigned int k = lid; k < MIO_BN_LESS;
+#if(MIO_LAYOUT_NHWC == 1 || MIO_BN_REM == 0)
+    __attribute__((opencl_unroll_hint(2))) for(unsigned int k = lid; k <
+#if MIO_LAYOUT_NHWC == 1
+                                                                     MIO_BN_NHW
+#else
+                                                                     MIO_BN_LESS
+#endif
+                                               ;
                                                k += MIO_BN_GRP0)
```
Comment (DrizztDoUrden) [inline src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:361]:
```suggestion
    const unsigned int k_limit =
#if MIO_LAYOUT_NHWC == 1
                                                                     MIO_BN_NHW;
#else
                                                                     MIO_BN_LESS;
#endif
    __attribute__((opencl_unroll_hint(2))) for(unsigned int k = lid; k < k_limit; k += MIO_BN_GRP0)
```

**Review by DrizztDoUrden — DISMISSED** [review]

**Comment by junliume** [comment]
ping @ce1adon for updates on this PR. Thanks!

**Comment by junliume** [comment]
ping @ce1adon for updates on this PR. Thanks!

**Comment by ce1adon** [comment]
> ping @ce1adon for updates on this PR. Thanks!

Debugging accuracy issues in ctest.

File: src/kernels/MIOpenBatchNormFwdTrainSpatial.cl (hunk 2 above)
Comment (junliume) [reply src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:361]: @DrizztDoUrden is this change critical? How about we maintain it as-is?

Comment (junliume) [reply src/kernels/MIOpenBatchNormFwdTrainSpatial.cl]: recommendations adopted

File: src/solver/batchnorm/forward_spatial_single.cpp [hunk 3]
```diff
@@ -129,54 +129,57 @@ BnFwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
...
-        ylocalsize   = 1024;
-        auto segment = int(std::ceil(double(in_cstride) / double(ylocalsize)));
-        xgridsize    = c;
-        ygridsize    = segment * ylocalsize;
-        ldsgcn       = ylocalsize / 64;
-        ldsnogcn     = ylocalsize;
-    }
-    // clang-format on
+            // clang-format off
+        if((in_nhw < 33554432 && in_cstride > 1024) ||
```
[the remark this answers is not in the document]
Comment (junliume) [reply src/solver/batchnorm/forward_spatial_single.cpp]: This part is repeated code from above, I suggest we do not change it.

File: src/kernels/MIOpenBatchNormBwdSpatial.cl [hunk 4]
```diff
@@ -229,9 +229,19 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
 
 #elif(MIO_BN_VARIANT == 1)
 
+#ifndef MIO_LAYOUT_NHWC
+#define MIO_LAYOUT_NHWC 0
+#endif
+
+#if MIO_LAYOUT_NHWC == 1
```
[the remark this answers is not in the document]
Comment (atamazov) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl:236]: Done everywhere but not here :smile:

File: src/kernels/MIOpenBatchNormBwdSpatial.cl [hunk 5]
```diff
@@ -229,9 +229,19 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
 
 #elif(MIO_BN_VARIANT == 1)
 
+#ifndef MIO_LAYOUT_NHWC
+#define MIO_LAYOUT_NHWC 0
+#endif
```
Comment (atamazov) [inline src/kernels/MIOpenBatchNormBwdSpatial.cl:234]: [Recommendation] Right now we define the default (0) but do not check if macro is correctly defined by MIOpen, so:
```suggestion
#ifndef MIO_LAYOUT_NHWC
#define MIO_LAYOUT_NHWC 0
#endif

#if (MIO_LAYOUT_NHWC != 0) || (MIO_LAYOUT_NHWC != 1)
#error MIO_LAYOUT_NHWC must be 0 or 1
#endif
```
or simply this (to enforce MIOpen to always correctly define this macro):
```suggestion
#if (MIO_LAYOUT_NHWC != 0) || (MIO_LAYOUT_NHWC != 1)
#error MIO_LAYOUT_NHWC must be 0 or 1
#endif
```
But this may be overkill of course :smile: 
If you like this, please make it everywhere, thanks.

**Review by atamazov — DISMISSED** [review]
No objections!

File: src/kernels/MIOpenBatchNormBwdSpatial.cl [hunk 6]
```diff
@@ -229,9 +229,23 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
...
+
+#if (MIO_LAYOUT_NHWC != 0) || (MIO_LAYOUT_NHWC != 1)
+#error MIO_LAYOUT_NHWC must be 0 or 1
```
Comment (atamazov) [inline src/kernels/MIOpenBatchNormBwdSpatial.cl:237]: [Recommendation] Oops, I am afraid we need double quotes around error message.
```suggestion
#error "MIO_LAYOUT_NHWC must be 0 or 1"
```

**Review by atamazov — APPROVED** [review]
LGTM

File: src/kernels/MIOpenBatchNormBwdSpatial.cl (hunk 6 above)
Comment (junliume) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl]: @atamazov I'm afraid I might have messed up something and trying to correct them:
```
[2021-12-16T21:18:03.373Z] 13/93 Test  #3: test_bn_3d_spatial_test ...............................***Failed  Error regular expression found in output. Regex=[FAILED]  1.34 sec
[2021-12-16T21:18:03.373Z] Choosing smaller input values for low dims
[2021-12-16T21:18:03.373Z] MIOpen(OpenCL): Error [BuildProgram] Build log: /tmp/comgr-547839/input/CompileSource:446:2: error: "MIO_LAYOUT_NHWC must be 0 or 1"
[2021-12-16T21:18:03.373Z] #error "MIO_LAYOUT_NHWC must be 0 or 1"
[2021-12-16T21:18:03.373Z]  ^
[2021-12-16T21:18:03.373Z] 1 error generated.
[2021-12-16T21:18:03.373Z] Error: Failed to compile opencl source (from CL or HIP source to LLVM IR).
```

Comment (junliume) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl]: Maybe #error macro is not accepted in COMGR or hipRTC?

Comment (atamazov) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl]: OMG we shall use `&&`, not `||` 🤕

Comment (junliume) [reply src/kernels/MIOpenBatchNormBwdSpatial.cl]: @atamazov ah, I'm so distracted and should have figured it out sooner :) I'm going to change it really soon.

**Comment by atamazov** [comment]
@junliume 
> Error fix f85cedd 

:+1: :+1: :+1:

**Review by atamazov — APPROVED** [review]
LGTM

**Comment by atamazov** [comment]
@ce1adon If would be nice if you bless this PR as well ;)

**Comment by atamazov** [comment]
@junliume The CI failure at run 19 was due to #1351

**Review by junliume — APPROVED** [review]
